### PR TITLE
CI: Gather test results

### PR DESCRIPTION
**Pull request by masterleinad**
This pull request makes the unit test results visible in the respective sections for the `jenkins` and `GItLab` CI.

Examples in https://cloud.cees.ornl.gov/jenkins-ci/job/Kokkos/job/PR-7470/8/testReport/ and https://gitlab.spack.io/kokkos/kokkos/-/pipelines/863294/test_report?job_name=NVIDIA-GH200.

### CODE REVIEW

**Comment by masterleinad** [comment]
> For the gitlab ci I do see a Tests tab with statistics, that said the log output has this:

https://gitlab.spack.io/kokkos/kokkos/-/pipelines/862573/test_report looks fine to me. No idea why it needed three attempts to succeed uploading the artifact. The Intel GPU build succeeded on the first try.

**Comment by crtrott** [comment]
Question: what do we gain with this? I looked at the test overview thing (in both cases) and its not clear that it would give you an easy way to get to the test-output in case of failure (nothing failed so maybe that is why?). I.e. if it only gives you a number or even a list of failed tests, without being able to link back to the output of the test I am not sure that we gain anything useful, since you would still need to click on the raw log file and scroll through it to see error report.

**Comment by masterleinad** [comment]
https://gitlab.spack.io/kokkos/kokkos/-/pipelines/863294/test_report?job_name=NVIDIA-GH200 has useful results and I'm used to seeing similar output for jenkins but that currently doesn't look quite right (https://cloud.cees.ornl.gov/jenkins-ci/job/Kokkos/job/PR-7470/7/testReport/). On the other hand, even having the list of failed tests there is more useful than needing to go through all the logs in my opinion.

**Comment by masterleinad** [comment]
Now, the test reports on `jenkins` are also good, see https://cloud.cees.ornl.gov/jenkins-ci/job/Kokkos/job/PR-7470/8/testReport/.

**Review by crtrott — APPROVED** [review]
Yeah that seems like useful output now :-)

**Review by Rombur — APPROVED** [review]

File: .jenkins [hunk 1]
```diff
@@ -91,11 +97,12 @@ pipeline {
                                 -DKokkos_ENABLE_BENCHMARKS=ON \
                                 -DKokkos_ENABLE_HIP=ON \
                               .. && \
-                              make -j8 && ctest --verbose'''
+                              make -j8 && ctest --no-compress-output -T Test --verbose'''
```
Comment (dalg24) [inline .jenkins:100]: Did you look at the size of these XML files that are produced as compared to the size of the logs that are uploaded to the server?
Should we be worried about disk space?

Comment (masterleinad) [reply .jenkins:100]: The `xml` files are between 56 KiB and 128 KiB per report:
- https://cloud.cees.ornl.gov/jenkins-ci/job/Kokkos/job/PR-7470/9/execution/node/179/ws/build/Testing/20241022-1416/ 56.07 KiB
- https://cloud.cees.ornl.gov/jenkins-ci/job/Kokkos/job/PR-7470/9/execution/node/48/ws/build/Testing/20241022-1330/ 104.55 KiB
- https://cloud.cees.ornl.gov/jenkins-ci/job/Kokkos/job/PR-7470/9/execution/node/180/ws/build/Testing/20241022-1511/ 127.70 KiB
- https://cloud.cees.ornl.gov/jenkins-ci/job/Kokkos/job/PR-7470/9/execution/node/183/ws/build/Testing/20241022-1423/ 117.89 KiB
- https://cloud.cees.ornl.gov/jenkins-ci/job/Kokkos/job/PR-7470/9/execution/node/184/ws/build/Testing/20241022-1404/ 111.32 KiB
- https://cloud.cees.ornl.gov/jenkins-ci/job/Kokkos/job/PR-7470/9/execution/node/185/ws/build/Testing/20241022-1435/ 113.52 KiB

That's negligible compared with the size of the build directories, KokkosCore_UnitTest_Cuda alone is 154.53 MiB in https://cloud.cees.ornl.gov/jenkins-ci/job/Kokkos/job/PR-7470/9/execution/node/186/ws/build/core/unit_test.

The size of the whole test output for that build is 4.3 MiB

File: .jenkins [hunk 2]
```diff
@@ -62,7 +62,13 @@ pipeline {
                                 -DKokkos_ENABLE_LIBQUADMATH=ON \
                                 -DKokkos_ENABLE_SERIAL=ON \
                               .. && \
-                              make -j8 && ctest --verbose && gcc -I$PWD/../core/src/ ../core/unit_test/tools/TestCInterface.c'''
+                              make -j8 && ctest --no-compress-output -T Test --verbose && gcc -I$PWD/../core/src/ ../core/unit_test/tools/TestCInterface.c'''
+                    }
+                    post {
+                        always {
+                            sh 'ccache --show-stats'
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply .jenkins]: Fixed. There are still various builds where we zero stats and reports stats that don't use `ccache`. If we think it matters, we can fix elsewhere.

File: .jenkins (hunk 1 above)
Comment (dalg24) [reply .jenkins:99]: > negligible compared with the size of the build directories

Build directories are not uploaded to the server.  I suspect these xml files are.

Comment (masterleinad) [reply .jenkins:99]: So we are talking about a maximum of 128*10 KiB or so in `xml` files which is around a quarter of the log file.

Comment (dalg24) [reply .jenkins:99]: Should we trade --verbose for --output-on-failure?
One downside is we would not be able to follow progress on a single test :/
What do you think?

**Review by dalg24 — COMMENTED** [review]
I am fine to merge as is.  You might want to investigate using the Jenkins JUnit support to simplify the configuration

**Pull request merged**